### PR TITLE
check grant type of client requesting code or access token

### PR DESCRIPTION
**Pull request by vinebrancho**
Authentication server needs to check grant type of client when it request authorization code or access token.
if instance of 'Application' model for the client has 'grantTypes' property, authentication server check if requested type is allowed.

Additionally, I add another patch checking user data regarding email verification.

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

File: lib/oauth2-loopback.js [hunk 1]
```diff
@@ -44,6 +44,24 @@ function isExpired(tokenOrCode) {
...
 }
 
+function isAllowedGrantType(client, requestedType) {
+  if (!client.grantTypes) {
+    // this case allows all types
+    return true;
+  }
+
+  var allowed = false;
+  client.grantTypes.forEach(function (type) {
```
Comment (raymondfeng) [inline lib/oauth2-loopback.js]: Let's use plain `for` loop to avoid the performance penalty.

**Comment by raymondfeng** [comment]
BTW, we should extend to check `scopes', 'tokenTypes' etc.

**Comment by raymondfeng** [comment]
I'm adding more checks to https://github.com/strongloop/loopback-component-oauth2/pull/18

**Pull request closed without merging**